Improve logs

luckysori opened the pull request:
- Add a logger to control log levels.
- Use nice colours.
- Format according to #4.
- Read optional `--loglevel <error | info | data | verbose>` from command line to set log level. Defaults to log level `info` if  nothing is passed.
- Improve logs in general. This is subjective and is influenced by #4's suggestions. I'm happy to change things.

## Examples

- Info:

![Oct03::084240](https://user-images.githubusercontent.com/9418575/66087573-4fae9f80-e5bb-11e9-8e2e-7a6aef4e2ad3.png)

- Data:

![Oct03::085323](https://user-images.githubusercontent.com/9418575/66087608-6ce36e00-e5bb-11e9-9926-a9bb586cb28f.png)

---

Fixes #4.
Fixes #9.

luckysori committed:
Use custom logger

luckysori committed:
Log balances in order on exit

luckysori committed:
Format logs to use single colour per line

Also, the format is now `[<msg.level>] <msg.message>`.

luckysori committed:
Format log messages about offers similarly

luckysori committed:
Improve comment

luckysori committed:
Log wallet balances as floats with 2 decimal places

luckysori committed:
Move logger instantiation to bottom of index.ts

luckysori committed:
Check for env file at the start of main

luckysori committed:
Move logger to separate file

luckysori committed:
Instantiate logger on demand instead of passing it around

luckysori committed:
Log error and exit where the logger is initialised

luckysori committed:
Print balances on the same line

luckysori committed:
Format data log messages differently

luckysori committed:
Move away from swap accept and decline in logs

luckysori committed:
Always mention sell asset/coin, then buy asset/coin

luckysori committed:
Remove the bad log

We do lose the ability to log on what blockchain the action is being
performed but that can be deduced or reintroduced eventually.

luckysori committed:
Remove unused dev dependency

luckysori committed:
Remove full stop

D4nte commented on src/helloSwap.ts:
```diff
@@ -97,24 +101,36 @@ export class HelloSwap {
+                                    `Cancelled swap with ID: ${simpleSwap.id}`
+                                );
+                                this.logger.data(
+                                    `with parameters: sell ${JSON.stringify(
```
I think in Javascript you can implement a `display` function that is used when doing display things to avoid doing `JSON.stringify` inline. It may be interesting to look into it.

luckysori replied:
Briefly looked into this, but couldn't find anything. Would be better though!

thomaseizinger replied:
We could also use format strings: "sell %s", that is very common with loggers.

luckysori replied:
Oh yeah, I think I know how to do this now with `winston`.

luckysori replied:
`JSON.stringify` is no more!

D4nte commented on src/logger.ts:
```diff
@@ -0,0 +1,66 @@
+    }).loglevel;
+
+    if (
+        level !== "error" &&
```
Are we actually using that many levels?

luckysori replied:
Yes, but verbose can go away because it's only used once.

D4nte replied:
Aiming for simple so I'd be keen to only have 3 levels at this stage:
- normal (info)
- With blockchain traces (trace or data)
- error (for errors :])

luckysori replied:
Gonna remove the ability to set it to `verbose` _and_ `error`, because if you set it to `error` it's actually not even useful (you don't know when the swap has happened).

luckysori committed:
Simplify log levels

thomaseizinger commented on src/logger.ts:
```diff
@@ -0,0 +1,66 @@
+});
+
+export function createLogger() {
+    const level = minimist(process.argv.slice(2), {
```
I'd happier if `process.XYZ` stuff happens only on the very root of the application to avoid implicit dependencies on global variables.

Can be a follow-up ticket :)

luckysori replied:
I tend to agree, but I couldn't figure out another way to avoid passing either the logging level or the logger around.

thomaseizinger replied:
I'd opt for passing around the log-level. Basically extract `process.argv.slice(2)` into a parameter would already do it :)

thomaseizinger commented on src/logger.ts:
```diff
@@ -0,0 +1,55 @@
+    return winston.createLogger({
+        levels: {
+            error: 0,
+            maker: 1,
```
Instead of abusing log levels here, I'd suggest to use customize the `format` based on a parameter that was passed in. You can then create an instance specific logger for maker and taker and don't have to pass it on every invocation.

It would also have the advantage of maker and taker being able to log on different levels :)
Can do this together in a follow-up PR :)

luckysori replied:
It's probably abusing log levels, yep. Probably best to deal with it in a follow-up PR as you said :relaxed:

thomaseizinger reviewed (approved):
Two things noted, happy to postpone to follow-up PRs :)

da-kami reviewed (approved):
like the adapted output!

luckysori committed:
Use string interpolation for logs

luckysori committed:
Remove logs related to the exchange of protocol parameters

Also:

- Remove logger from main and just instantiate it when we log the
  wallet balances.
- `await` on logging the wallet balances on `hello-swap` app startup
  so that logs appear in the right order.

luckysori committed:
Merge branch 'master' into 4-log-levels
